cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -176,19 +177,65 @@ mkTracers traceOptions tracer = Tracers
-                                         meta <- mkLOMeta Debug Confidential
-                                         let tr' = appendName "slotNum" tr
-                                         traceNamedObject tr' (meta, logValue)
+                  ChainDB.SwitchedToChain _ c ->
+                    when (slots c > 0) $ do
+                      meta <- mkLOMeta Notice Public
+                      let tr' = appendName "metrics" tr
+                          sNum :: Integer = fromIntegral $ slotN c
+                          bNum :: Integer = fromIntegral $ blockN c
+                          sl = slots c
+                          bl = blocks c
+                          density = toRational bl / toRational sl
+                          epochSlots :: Integer = 21600  -- TODO
+                      traceNamedObject tr (meta, LogMessage $ "density based on blocks:" <> pack (show bl) <> " slots:" <> pack (show sl))
+                      traceNamedObject tr' (meta, LogValue "density" . PureD $ fromRational density)
+                      traceNamedObject tr' (meta, LogValue "slotNum" . PureI $ sNum)
+                      traceNamedObject tr' (meta, LogValue "blockNum" . PureI $ bNum)
+                      traceNamedObject tr' (meta, LogValue "slotInEpoch" . PureI $ sNum `rem` epochSlots)
+                      traceNamedObject tr' (meta, LogValue "epoch" . PureI $ (sNum `div` epochSlots))
                   _ -> pure ()
               _ -> pure ()
+      where
+        slotN :: forall block . AF.HasHeader block
+              => AF.AnchoredFragment block -> Word64
+        slotN frag  = unSlotNo $ fromWithOrigin 0 (AF.headSlot frag)
+        -- Slot of the tip - slot @k@ blocks back. Use 0 as the slot for genesis
+        -- includes EBBs
+        slots :: forall block . AF.HasHeader block
+              => AF.AnchoredFragment block -> Word64
+        slots frag  = slotN frag
+                    - unSlotNo (fromWithOrigin 0 (pointSlot (AF.anchorPoint frag)))
+        blocks :: forall block . AF.HasHeader block
+               => AF.AnchoredFragment block -> Word64
+        -- Block numbers start at 1. We ignore the genesis EBB, which has block number 0.
+        blocks frag = blockN frag - firstBlock frag
+        blockN :: forall block . AF.HasHeader block
+                  => AF.AnchoredFragment block -> Word64
+        blockN frag = unBlockNo $ fromMaybe 1 (AF.headBlockNo frag)
+        firstBlock :: forall block . AF.HasHeader block
+                   => AF.AnchoredFragment block -> Word64
+        firstBlock frag = case unBlockNo . blockNo <$> AF.last frag of
+                  -- Empty fragment, no blocks. We have that @blocks = 1 - 1 = 0@
+                  Left _ -> 1
+                  -- The oldest block is the genesis EBB with block number 0,
+                  -- don't let it contribute to the number of blocks
+                  Right 0 -> 1
+                  Right b -> b
```

Comment:
I would be inclined to put this in a nice self-contained function: `chainDensity :: HasHeader blk => AnchoredFragment blk -> Rational`. I see you're also tracing most intermediary values, which complicates this.

Reply:
yes, we need access to those values for display and export to monitoring.

Reply:
The intermediate values you're tracing are just the slots and blocks, so return them too:
```
chainDensity :: HasHeader blk => AnchoredFragment blk -> (Word64, Word64, Rational)
```
So you can pull it out into a function.

Reply:
done